nvs: fix initialization of fs->data_wra on alignment change

Damian-Nordic opened the pull request:
When existing records stored in the NVS are not properly
    aligned according to the current flash driver requirements,
    fs->data_wra may be initialized with an unaligned address.
    Fix the initialization code, so that fs->data_wra is rounded
    up to the nearest multiple of the current flash driver block
    size.
    
    The situation may occur during a firmware upgrade which
    introduces a new flash driver or changes its parameters.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

Damian-Nordic committed:
nvs: fix initialization of fs->data_wra on alignment change

When existing records stored in the NVS are not properly
aligned according to the current flash driver requirements,
fs->data_wra may be initialized with an unaligned address.
Fix the initialization code, so that fs->data_wra is rounded
up to the nearest multiple of the current flash driver block
size.

The situation may occur during a firmware upgrade which
introduces a new flash driver or changes its parameters.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

Laczen commented:
@Damian-Nordic, thanks for the PR.

Would it be possible to add this in a alignment check routine that walks over all ate's to see if any `ate.offset` points to an unaligned location ? This way unaligned data in all sectors would be detected.

de-nordic commented:
> 
> 
> @Damian-Nordic, thanks for the PR.
> 
> Would it be possible to add this in an alignment check routine that walks over all ate's to see if any `ate.offset` points to an unaligned location? This way unaligned data in any sector would be detected.

@Laczen what do you think off providing some recovery mechanism for that issue, would you be able to propose some solution?

This detects a case when data is correct and it is still there and is readable, although the problem is that the reading mechanism will either read every N'th ate (N = new_alignment / old_alignment) when write alignment have increased  or will find gaps if write alignment have decreased.
It should be possible to figure out how the write alignment changed and rewrite data with new alignment.

de-nordic commented on subsys/fs/nvs/nvs.c:
```diff
@@ -635,6 +635,14 @@ static int nvs_startup(struct nvs_fs *fs)
+			/* check if existing NVS entry is properly aligned according to
+			 * the current flash driver requirements
+			 */
+			if (fs->data_wra & (fs->flash_parameters->write_block_size - 1)) {
```
There should probably be some `LOG_ERR` here as with the other `-EINVAL` (in `nvs_init`)

Laczen commented:
@Damian-Nordic, @de-nordic, been thinking about this a little more, should this even raise an error. If data_wra is adjusted to an aligned position at the end of init it does not matter at all.

Damian-Nordic commented:
@Laczen Currently, `data_wra` is not adjusted to an aligned position, but I guess it can be fixed easily. Does unaligned reads never generate any issues? In my case aligning `data_wra` up should fix the problem, but I'm thinking about the general case.

Laczen commented:
> @Laczen Currently, `data_wra` is not adjusted to an aligned position, but I guess it can be fixed easily. Do unaligned reads never generate any issues? In my case aligning `data_wra` up should fix the problem, but I'm thinking about the general case.

Just added something above.

Damian-Nordic commented:
@Laczen I'll test that fix and let you know. Thanks.

de-nordic commented:
> 
> 
> @Damian-Nordic, @de-nordic, been thinking about this a little more, should this even raise an error. If data_wra is adjusted to an aligned position at the end of init it does not matter at all.

@Laczen what about the closed ate detection loop in `nvs_startup`?
Just before that loop the `ate_size`, which is used to figure address of the last ate in sector, is aligned to *the current* write sector; if you assume that for example write alignment doubled, this means that it will start read at the second ate from the end of sector and that may contain ate followed by "not-closed" ate but will be detected as closing ate as a whole;  such situation may lead to detecting all sectors as closed.

Laczen commented:
@de-nordic, i didn't consider an ate size change (this would only happen when changing from <=4 byte alignment to 8). Then none of the ate's would be valid (crc8 is in the wrong place).

de-nordic commented:
> 
> 
> @de-nordic, i didn't consider an ate size change (this would only happen when changing from <=4 byte alignment to 8 or vice versa). Then none of the ate's would be valid (crc8 is in the wrong place).

I think it might be detected as ok. The `nvs_ate_crc8_check` is calculating crc on ate entry taking the size of `nvs_ate` structure minus the size of crc:
`crc8 = crc8_ccitt(0xff, entry, offsetof(struct nvs_ate, crc8))`
the update of crc is done with `nvs_ate_crc8_update` that does:
```
crc8 = crc8_ccitt(0xff, entry, offsetof(struct nvs_ate, crc8));
entry->crc8 = crc8;
```
So it operates within `nvs_ate` structure that is then directly written/read at the write_block alignment.
This means that the write block is N times greater then the size of `nvs_ate`  the `nvs_ate` at the beginning of the block, if there is one, will be checked and detected to  be ok.

Damian-Nordic commented:
@Laczen @de-nordic 
I tested the `fs->data_wra += nvs_al_size(fs, last_ate.offset + last_ate.len);` code and it works in my case, so I updated the PR accordingly. 
I understand what you're saying about switching from `<=8` to `>=16` alignment and silently skipping `ate`s in such a case, but I'm not sure what we can do about it. Probably the whole NVS would need to have kind of signature/header written to flash, in which size of the `ate` structure would be included, but I guess there's currently no such thing.

de-nordic commented:
>
> I tested the `fs->data_wra += nvs_al_size(fs, last_ate.offset + last_ate.len);` code and it works in my case, so I updated the PR accordingly.

If @Laczen approves the change I will approved but I still think there will be problem in such case:

let N be byte write bloc prior to upgrade,let  M number of sectors:
1) use N byte write block to fill and close 1 to M - 1 sectors and write one ate in M'th sector;
2) perform upgrade via flashing new image that uses 2*N write block;
3) boot the device;

In such case the nvs_startup will at 3) will detect all sectors to be closed and return with `-EDEDLK`.
@Laczen  can you verify me here?

Laczen commented:
> > I tested the `fs->data_wra += nvs_al_size(fs, last_ate.offset + last_ate.len);` code and it works in my case, so I updated the PR accordingly.
> 
> If @Laczen approves the change I will approved but I still think there will be problem in such case:
> 
> let N be byte write bloc prior to upgrade,let M number of sectors:
> 
> 1. use N byte write block to fill and close 1 to M - 1 sectors and write one ate in M'th sector;
> 2. perform upgrade via flashing new image that uses 2*N write block;
> 3. boot the device;
> 
> In such case the nvs_startup will at 3) will detect all sectors to be closed and return with `-EDEDLK`.
> @Laczen can you verify me here?

I think the change is about going from aligment 1 to alignment 4 or 8. These would be corrected.

The case of changing from 8 to 16 bytes alignment is indeed not handled.  I don't think your case is valid because there always is an empty sector and this is not influenced by the alignment. BTW, your case would have been good since it would block access. In reality a change in alignment from 8 to 16 bytes or vice versa will result in problems. This would be one of the things that could be corrected in nvs version 2 (it's already 3 years since it was introduced).

de-nordic commented:
> 
> I think the change is about going from aligment 1 to alignment 4 or 8. These would be corrected.

Yes, but the fix should be universal and take in to account the other increase scenarios. I think the data write is covered but ate writes are not
I am approving the fix for the data path but will be logging an issue for fixing the ate reads.

de-nordic reviewed (approved):
The PR fixes problem with data alignment but additional work on ate will be needed.

nvlsianpu commented:
^^ @de-nordic Can you open issue for tracking that work?

nvlsianpu reviewed (changes requested):
I think we need additional comment in the code which explain that this line allow compatibility between NVS memory images which have same physicals ate size but different write alignment.
Also need to add such record to the NVS documentation.

Damian-Nordic commented:
> I think we need additional comment in the code which explain that this line allow compatibility between NVS memory images which have same physicals ate size but different write alignment.
> Also need to add such record to the NVS documentation.

@nvlsianpu Do you think someone else could follow up on this? I just wanted to help fix the issue which I spotted during testing, but if you think it requires more changes I may not find time to work on that now.

nvlsianpu committed:
fs/nvs: document flash wbs migration capability

Document the fact than NVS image of different wbs
are compatible each other if psychical ATE size
is kept.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>
